app/src/main/java/ch/epfl/sweng/erpa/services/GCP/GCPRemoteServicesProvider.java

```diff
@@ -0,0 +1,140 @@
+
+    @Getter private final GCPGameService gameService;
+    @Getter private final GCPUserProfileService userProfileService;
+
+    public static final String API_URL = "https://erpa-sweng.appspot.com";
+
+    // Public for debugging purposes :)
+    public static Gson gson = new GsonBuilder()
+            .registerTypeAdapterFactory(OptionalTypeAdapter.FACTORY)
+            .registerTypeAdapterFactory(GameDifficultyTypeAdapter.FACTORY)
+            .create();
+
+    @Getter @NonNull private Retrofit retrofit = new Retrofit.Builder()
+            .baseUrl(API_URL)
+            .addConverterFactory(GsonConverterFactory.create(gson))
+            .build();
+
+
+    @Inject public GCPRemoteServicesProvider() {
+        gameService = new GCPGameService(this);
+        userProfileService = new GCPUserProfileService(this);
+    }
+
+    public <T> void executeRequest(Call<T> call, Consumer<T> callback, Consumer<Exception> exceptionConsumer) {
+        Callback<T> cb0 = new Callback<T>() {
+
+            @Override public void onResponse(@NonNull Call<T> call0, Response<T> response) {
+                if (response.isSuccessful() && response.body() != null) {
+                    callback.accept(response.body());
+                } else {
+                    try {
+                        callback.accept(null);
+                        Log.d("executeRequest", "Fetching game list failed", new ServerException(response.code(), response.errorBody().string()));
+                    } catch (IOException e) {
+                        // Ignored
+                    }
+                }
+            }
+
+            @Override public void onFailure(@NonNull Call<T> call0, Throwable t) {
+                exceptionConsumer.accept(transformThrowable(t));
+            }
+        };
+        call.enqueue(cb0);
+    }
+
+    private static Exception transformThrowable(Throwable t) {
+        if (Exception.class.isAssignableFrom(t.getClass())) {
+            return (Exception) t;
+        } else {
+            return new RuntimeException(t);
+        }
+    }
+
+    <T> Response<T> executeRequestBlocking(Call<T> call) throws IOException {
+        return call.execute();
+    }
+
+    <T> void executeAndThrowOn404(Call<T> call) throws IOException, ServerException {
+        Response<?> response = executeRequestBlocking(call);
+        if (response.code() == 404)
+            throw new NoSuchElementException("Server couldn't find user profile with this id");
+        if (!response.isSuccessful()) throw new ServerException();
+    }
+
+    /**
+     * converts a response to an object, and discards the error
+     *
+     * @param resp the response
+     * @return the optional, if the request was successful
+     */
+    private static <T> Optional<T> responseToOptional(Response<T> resp) throws ServerException, IOException {
+        T optT;
+        if (resp.isSuccessful() && (optT = resp.body()) != null) {
```

Comment:
Ugly, use map: `return Optional.of(rest).filter(...).map(...).peek(...maybethrow...)`